File: Easy_Problems/10. Minimum Waiting Time/Python/Solution1.py
# Greedy Approach
# Time Complexity: O(N · log(N)), where N is the number of queries
# Space Complexity: O(1)

def get_total_min_waiting_time(queries):
    queries.sort()
    total_waiting_time = 0
    next_waiting_time = 0
    
    for duration in queries:
        total_waiting_time += next_waiting_time
        next_waiting_time += duration
    
    return total_waiting_time

File: Easy_Problems/10. Minimum Waiting Time/Javascript/Solution1.js
// Greedy Approach
// Time Complexity: O(N · log(N)), where N is the number of queries
// Space Complexity: O(1)

function getTotalMinWaitingTime(queries) {
    queries.sort((a, b) => a - b);
  
    let nextWaitingTime = 0;
    let totalWaitingTime = 0;
  
    for (const duration of queries) {
      totalWaitingTime += nextWaitingTime;
      nextWaitingTime += duration;
    }
  
    return totalWaitingTime;
}


// Tests for getTotalMinWaitingTime
const testGetTotalMinWaitingTime = () => {
    console.log(getTotalMinWaitingTime([3, 2, 1, 2, 6]) == 17)
    console.log(getTotalMinWaitingTime([10]) == 0)
    console.log(getTotalMinWaitingTime([1, 10]) == 1)
}